### ISSUE TITLE
Nightly build has invalid semver

### ISSUE THREAD

**Issue by mhassan1** [yarnpkg/yarn#8534]
The nightly build of Yarn v1 has invalid semver (currently `yarn-v1.23.0-20210102.0010`). It is invalid because of the leading zero after the `.` near the end:

```js
require('semver').valid('1.23.0-20200928.0848') // -> null
require('semver').valid('1.23.0-20200928.1848') // -> "1.23.0-20200928.1848"
```

This seems to be caused by the generation of the version [here](https://github.com/yarnpkg/yarn/blob/14bcd15a401e063c2130ab31568c80aeb914165f/scripts/set-dev-version.js#L16):

```js
const formattedDate =
  date.getUTCFullYear() +
  leftPad(date.getUTCMonth() + 1) +
  leftPad(date.getUTCDate()) +
  '.' +
  leftPad(date.getUTCHours()) +
  leftPad(date.getUTCMinutes());
```

Is there any appetite for changing this so that it does not generate invalid semver? Perhaps using a `-` instead of a `.` before the time? Or prepending a non-zero before the hour? I don't know if there's a dependency on the current format.

In our case, this causes breakage to a Yarn v2 workflow that packs a GitHub dependency using Yarn Classic as part of the project installation.

**Comment by merceyz** [comment]
Closing as fixed in https://github.com/yarnpkg/yarn/commit/3119382885ea373d3c13d6a846de743eca8c914b

**Comment by dchambers** [comment]
@merceyz, could this change please be backported to a `2.4.2` release? Otherwise from `2.4.1` you have to temporarily `yarn set version from sources` before you can `yarn set version 3.0.0-rc.2`, and it won't be obvious to newcomers that this may work.

And that's not to mention that in my case I also had to `yarn plugin remove @yarnpkg/plugin-workspace-tools` before performing the version upgrade, and then re-issuing `yarn plugin import @yarnpkg/plugin-workspace-tools` once the version upgrade was complete.

**Comment by merceyz** [comment]
Did you comment on the wrong issue? This is for nightly releases of Yarn 1.x

> And that's not to mention that in my case I also had to yarn plugin remove @yarnpkg/plugin-workspace-tools before performing the version upgrade, and then re-issuing yarn plugin import @yarnpkg/plugin-workspace-tools once the version upgrade was complete.

We fixed that in https://github.com/yarnpkg/berry/pull/2848 so upgrading from 2.x to 3.0.0-rc.3 should not have that problem

**Comment by dchambers** [comment]
> Did you comment on the wrong issue? This is for nightly releases of Yarn 1.x
> 
> > And that's not to mention that in my case I also had to yarn plugin remove @yarnpkg/plugin-workspace-tools before performing the version upgrade, and then re-issuing yarn plugin import @yarnpkg/plugin-workspace-tools once the version upgrade was complete.
> 
> We fixed that in [yarnpkg/berry#2848](https://github.com/yarnpkg/berry/pull/2848) so upgrading from 2.x to 3.0.0-rc.3 should not have that problem

Ah, yes, good point! I think it may have been [this comment](https://github.com/yarnpkg/berry/issues/2321#issuecomment-753625556) which originally brought me across to this repo, but you're right that I'm definitely in the wrong place now :smile:. You can't do a direct upgrade from `2.4.1` to `3.0.0.rc.2` though (you get an `Invalid semver version` error), so I'll look to create an issue for that.

**Comment by merceyz** [comment]
> You can't do a direct upgrade from 2.4.1 to 3.0.0.rc.2 though (you get an Invalid semver version error), so I'll look to create an issue for that.

Assuming you meant `3.0.0-rc.2` then that's the issue we fixed in https://github.com/yarnpkg/berry/pull/2848

**Comment by dchambers** [comment]
> > You can't do a direct upgrade from 2.4.1 to 3.0.0.rc.2 though (you get an Invalid semver version error), so I'll look to create an issue for that.
> 
> Assuming you meant `3.0.0-rc.2` then that's the issue we fixed in [yarnpkg/berry#2848](https://github.com/yarnpkg/berry/pull/2848)

I see, got it! And I also see that that fix only got merged just 3 days ago, so won't have made it its way to a `2.x.y` release yet. Thanks for fixing this, and sorry to have bothered you again! :grinning:

**Comment by merceyz** [comment]
> so won't have made it its way to a 2.x.y release yet

It will be fixed with the 3.0.0-rc.3 release, the invalid semver error is actually that the 3.0.0-rc.2 bundle fails to load your 2.x plugins and prints an error which is not a valid semver version. https://github.com/yarnpkg/berry/pull/2853 prevents it from doing that while https://github.com/yarnpkg/berry/pull/2848 allows it to load 2.x plugins

**Comment by dchambers** [comment]
> > so won't have made it its way to a 2.x.y release yet
> 
> It will be fixed with the 3.0.0-rc.3 release, the invalid semver error is actually that the 3.0.0-rc.2 bundle fails to load your 2.x plugins and prints an error which is not a valid semver version. [yarnpkg/berry#2853](https://github.com/yarnpkg/berry/pull/2853) prevents it from doing that while [yarnpkg/berry#2848](https://github.com/yarnpkg/berry/pull/2848) allows it to load 2.x plugins

Got it! Thanks for clarifying :bow:
